### ISSUE TITLE
Konsole issues

### ISSUE THREAD

**Issue by dbfin** [romkatv/powerlevel10k#1051]
When I clear the KDE's Konsole screen via a menu or by pressing Ctrl+Shift+K, the Konsole clears the screen as expected, but leaves 2 lines instead of 1 at the top. The first line is usually the copy of the second one with the last symbol disappearing. It looks like this after the clear screen command:
```
~ ···································································································· 01:06:5
~ ··································································································· 01:06:52
❯ 
```
The expected behavior and look would be like this:
```
~ ··································································································· 01:06:52
❯ 
```

**Comment by romkatv** [comment]
Konsole doesn’t know where prompt is, so it may clear the wrong part. To avoid this issue, use Ctrl+L instead. This shortcut asks Zsh to clear the screen.

**Comment by romkatv** [comment]
I've debugged this a bit further and found out that my previous comment was incorrect. The reason why a prompt line gets duplicated when you press <kbd>Ctrl+Shift+K</kbd> in Konsole is a bug in Konsole or one of its dependencies. I've reported it here: https://bugs.kde.org/show_bug.cgi?id=427577.

This bug is not specific to Powerlevel10k or even to Zsh.

As a workaround you can add the following to `~/.zshrc`:

```zsh
function clear-screen-and-scrollback() { printf '\e[3J' >$TTY && zle clear-screen }
zle -N clear-screen-and-scrollback
bindkey '^[^k' clear-screen-and-scrollback
```

Now you can press <kbd>Ctrl+Alt+k</kbd> to clear the screen together with scrollback (<kbd>Ctrl+L</kbd> will clear the screen without clearing the scrollback, just like before). In addition to working correctly in Konsole, this solution will work in all terminals. This is unlike the <kbd>Ctrl+Shift+K</kbd> shortcut that only works in Konsole.

**Comment by dbfin** [comment]
Thanks for fast reply and doing all this!

I read their response, and cannot agree with them calling their dirty hack a brilliant one -- at least this is highly subjective. Instead of fixing the underlying issue, which is btw not present in other terminals, someone takes all pride of coming up with a hacky idea, and then nothing in the world can change their mind about it, which is a shame. As in the end, thanks to such hacks, we have to live with all this software filled with brilliant bugs.

As a side question. Suppose they indeed increase the size of the terminal and then decrease it to make sure that the prompt is redrawn, if I even understood that correctly ))). Maybe you can then react somehow on both signals (I presume there should be 2 resize signals then, no?)... Or are there any other difficulties? What is the difference between this and if I manually increase and then decrease the terminal width?

Anyway, thanks for the workaround as well. For now I just mapped `Ctrl+K` to `Clear scrollback`, and press `K` and `L` in a quick succession while holding `Ctrl`, which works as `Clear scrollback and reset` should work, and is even easier than `Ctrl+Shift+K` thanks to `K` and `L` being next to each other.

I will wait then till this issue is resolved one either side. As I can see by now, the Konsole's developers are the cause of this issue, but maybe there is also on your side you can further do in the theme itself to workaround this Konsole's approach.

**Comment by dbfin** [comment]
> When you manually increase the size of the terminal from 80 to 81 columns, Zsh will think the terminal is 81 columns wide and this will actually be true.
> When you press Ctrl+Shift+K in Konsole, Konsole lies to Zsh. It makes Zsh believe that the terminal is 81 columns when it's actually only 80. This is the difference.

Yes, I figured that part, I guess I was asking doesn't Konsole then tells Zsh that its size is back to 80? This is what I meant asking whether Konsole sends 2 signals, a lie about the fake width 81 and then the truth about the actual width 80. I believe Konsole must do that so that the latest information `zsh` gets is the true width 80.

> As I mentioned earlier, Powerlevel10k has nothing to do with this. The hack in Konsole breaks Zsh.

Yes, this is my understanding at the moment as well.

> Have you seen the workaround I posted above? It has no downsides. It works perfectly in all terminals including Konsole. I'll repeat it here with a different binding in case you were confused by Ctrl+Alt+k.
> ...

No, I got that as well, thanks again. I will add something like this later to a script I have that is called from different terminals and contains some common initialization stuff. I was just lazy for now to do it your (proper) way... ))) So I just assigned a single shortcut for now instead, and since the keys `K` and `L` are together, in fact, I just hold `Ctrl` and press both of them together simultaneously, and it works the same way as `Ctrl+Shift+K` used to work.

I need to finish figuring out all the details of moving from p9k to p10k for now. Btw, using the moment, thanks for this great stuff you do! Very useful indeed! I used to have my own custom `PS1` with some Git integration, but of course no comparison to what you guys have implemented.

**Comment by romkatv** [comment]
> Yes, I figured that part, I guess I was asking doesn't Konsole then tells Zsh that its size is back to 80? This is what I meant asking whether Konsole sends 2 signals, a lie about the fake width 81 and then the truth about the actual width 80. I believe Konsole must do that so that the latest information `zsh` gets is the true width 80.

Yes, Konsole does indeed send another notification to Zsh informing it about the correct terminal dimensions. This is why you see correct prompt printed over the incorrect one. There is no way to undo the damage done by the first invalid notification, so the remains of the broken prompt (it's broken because Konsole lied about the terminal dimensions) remain.

> I need to finish figuring out all the details of moving from p9k to p10k for now. Btw, using the moment, thanks for this great stuff you do! Very useful indeed!

You are most welcome.

**Comment by dbfin** [comment]
So, I played with it a bit more, and thought that this double line effect should be caused by the extra character, and indeed it is, so here is another workaround: 
```
export ZLE_RPROMPT_INDENT=1
```
In fact, I believe this is the default value, so many may not even experience the bug I describe by default. The drawback of this default value is that there is a 1 character space on the RHS of the prompt. And this is why I changed the default value to 0, so the prompt takes the full width, and this (together with their approach to redraw things), as I understand now, caused the double line bug.

So, another solution, if someone encounters the same bug, is just leave this constant default, and live with the empty space on the RHS of the prompt. But I personally will probably add the universal approach you suggested instead. Though I will probably bind it to `Ctrl+L` instead.

---

On a side note, really love the way the config now works in p10k! Instead of about 70 customization lines in my p9k config, all I now have is roughly this:
```
typeset -g POWERLEVEL9K_HOME_FOLDER_ABBREVIATION='\UF015'
typeset -g POWERLEVEL9K_SHORTEN_DELIMITER='…'
typeset -g POWERLEVEL9K_DIR_FOREGROUND=7
typeset -g POWERLEVEL9K_DIR_SHORTENED_FOREGROUND=$POWERLEVEL9K_DIR_FOREGROUND
typeset -g POWERLEVEL9K_DIR_ANCHOR_FOREGROUND=15
typeset -g POWERLEVEL9K_DIR_ANCHOR_BOLD=false
typeset -g POWERLEVEL9K_BACKGROUND_JOBS_VERBOSE=true
```
And the rest is written to `.p10k.zsh` automatically by the customization wizard. I think this is a really great idea and approach.

**Comment by dbfin** [comment]
Yet another solution would be to convince Konsole's people to at least fake-shrink the terminal window instead of fake-enlarging it. This will probably resolve the double line issue, though now instead of two lines it will be leaving an extra space on the RHS of one line after `Ctrl+Shift+K`. But this is more acceptable (after all with the default `ZLE_RPROMPT_INDENT=1` you always have that space). And if you think about it, when they committed to this fake-size hack, they should have implemented it this way from the beginning, as lying to zsh that it has less space than in reality at least ensures zsh never tries to use more space than available, otherwise you have problems like we have, because zsh tries to use some space that is simply not there.

In any case, I feel like this issue can be closed now. We have figured out the cause of the issue, blamed others ))), and came up with at least 2 methods to workaround it. I don't think there is anything else you can do about it currently, except if you would agree with what I say above, and try to convince them to shrink the window instead of enlarging it. But in my opinion this would not solve the underlying fundamental problem with their design and approach. 

P.S. I personally returned `ZLE_RPROMPT_INDENT=0` as I did not like that extra space on the RHS, and just added the `Ctrl+L` binding to a function you suggested, and it works flawlessly for me. So, it is `Ctrl+L` for me from now on then.

**Comment by romkatv** [comment]
> But I personally will probably add the universal approach you suggested instead.

Good call. As I mentioned above, this solution has no downsides as far as I can tell. There is no need to search for workarounds given that a few lines of Zsh do the trick.

> In any case, I feel like this issue can be closed now.

Ditto. Closing.

**Comment by dbfin** [comment]
BTW, I just encountered a different problem but still related to setting `ZLE_RPROMPT_INDENT=0`: on Debian 10 my `zsh` setup was producing a prompt with the cursor right next to `>`, and then it would jump when trying to use autocompletion etc. I am pretty sure you know what I am talking about. So, I compared the versions, and while Arch of course moved to the latest `zsh` 5.8, Debian 10 has frozen it at version 5.7. So, my guess was that there is some bug fixed between these versions, as even disabling `powerlevel10k` completely or using `powerlevel9k` instead would not resolve the issue when setting `PS1` and `RPS1` together with `ZLE_RPROMPT_INDENT=0` manually.

But I also searched your issues just in case, and indeed found this: https://github.com/romkatv/powerlevel10k/issues/458#issuecomment-581516704. So, it seems I was correct that the issue is core and was fixed in `zsh` 5.8 only, but it also turned out that you are actually the one who fixed it. So, thanks for that!

For those encountering the same issue, I added this to `.zshrc` now:
```sh
autoload -U is-at-least
is-at-least 5.8 && export ZLE_RPROMPT_INDENT=0
```
This ensures that you won't get any problems caused by `zsh` itself related to setting this parameter in `zsh` prior to version 5.8.

**Comment by romkatv** [comment]
There were many bugs in Zsh that affected `ZLE_RPROMPT_INDENT=0`. 5.8 is the first version without such bugs.

When Powerlevel10k sees that you've enabled `ZLE_RPROMPT_INDENT=0` for a buggy version of zsh, it employs various workarounds (different for different versions), so that in most cases everything still works fine. It's not possible to apply these workarounds in 100% of cases though.

**Comment by dbfin** [comment]
Well, at least in Debian's Konsole 18 running Zsh 5.7 the problem is present, so I had to just switch it off -- that extra space on the right side is no issue, of course, but the jumping cursor is. In Arch's Konsole 20 running Zsh 5.8 no problems at all. So, thanks again for looking into this at the time, and fixing it!

**Comment by romkatv** [comment]
> Well, at least in Debian's Konsole 18 running Zsh 5.7 the problem is present

As I mentioned, all bugs have been fixed only in 5.8.

In 5.7 Powerlevel10k employs workarounds but they don't work for 100% of prompt configurations. They do work for most configurations though.

> In Arch's Konsole 20 running Zsh 5.8 no problems at all. So, thanks again for looking into this at the time, and fixing it!

It doesn't matter what terminal you are using. What matter is that the version of zsh is 5.8.

**Comment by dbfin** [comment]
> > Well, at least in Debian's Konsole 18 running Zsh 5.7 the problem is present

> As I mentioned, all bugs have been fixed only in 5.8.
> In 5.7 Powerlevel10k employs workarounds but they don't work for 100% of prompt configurations. They do work for most configurations though.

Right. This is why I mentioned that this is in Konsole, and also not in the latest version. I have not tested say Konsole 20 + Zsh 5.7 -- maybe your workarounds work there, who knows.

> > In Arch's Konsole 20 running Zsh 5.8 no problems at all. So, thanks again for looking into this at the time, and fixing it!

> It doesn't matter what terminal you are using. What matter is that the version of zsh is 5.8.

I cannot agree with that. After all, even this issue I started initially turned out to be about a bug that you can see in Konsole only due to their "brilliant hack". :) So, there still might be related bugs present in Konsole only that are not in any other terminal. In any case, it does not hurt to mention what terminal and which version you use.
